libantworld/agent.h

```diff
@@ -18,88 +18,65 @@ using namespace BoBRobotics::Pose;
 
 namespace BoBRobotics {
 namespace AntWorld {
-template<class LengthUnit = meter_t, class AngleUnit = radian_t>
 class AntAgent
-  : Positionable<LengthUnit>
-  , Rotatable<AngleUnit>
-  , Video::OpenGL
+  : Video::OpenGL
 {
 public:
-    AntAgent(GLFWwindow *window, Renderer &renderer, GLsizei readWidth, GLsizei readHeight);
-    Vector<LengthUnit> &getPosition() override;
-    Vector<AngleUnit> &getAttitude() override;
-    void setPosition(LengthUnit x, LengthUnit y, LengthUnit z) override;
-    void setAttitude(AngleUnit yaw, AngleUnit pitch, AngleUnit roll) override;
-    bool readFrame(cv::Mat &frame) override;
+    AntAgent(GLFWwindow *window, Renderer &renderer, GLsizei readWidth, GLsizei readHeight)
+      : Video::OpenGL(0, 0, readWidth, readHeight)
+      , m_Renderer(renderer)
+      , m_Window(window)
+    {}
+
+    template<class LengthUnit = meter_t>
+    Triple<LengthUnit> getPosition() const
+    {
+        return makeUnitTriple<LengthUnit>(m_Position);
+    }
+
+    template<class AngleUnit = degree_t>
+    Triple<AngleUnit> getAttitude() const
+    {
+        return makeUnitTriple<AngleUnit>(m_Attitude);
+    }
+
+    void setPosition(meter_t x, meter_t y, meter_t z)
+    {
+        m_Position[0] = x;
+        m_Position[1] = y;
+        m_Position[2] = z;
+    }
+
+    void setAttitude(degree_t yaw, degree_t pitch, degree_t roll)
+    {
+        m_Attitude[0] = yaw;
+        m_Attitude[1] = pitch;
+        m_Attitude[2] = roll;
+    }
+
+    bool readFrame(cv::Mat &frame)
+    {
+        // Clear colour and depth buffer
+        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
```

Comment:
I get the desire to do this but it does make the whole thing super inflexible - if you look at how antworld (which should be renamed) uses this stuff you wouldn't be able to use this class.

Reply:
I agree that it does make the class inflexible, but I'm not sure what the best approach is here. I have just moved the rendering stuff out of that function, but in retrospect I think it should probably stay in. Essentially what I was trying to do here is make something a bit like our current (somewhat crappy) Matlab scripts for achieving the same thing, where you can set an agent's position and get its current panoramic view.

Then we could write functions like this silly example:
``
template<class T>
void getViewFromOrigin(T &agent, cv::Mat &frame)
{
    agent.setPosition(0_m, 0_m, 0_m);
    agent.readFrame(frame);
}
``

The problem is that if we need to do extra things in between setting the position and reading the view then it won't be easy to write e.g. tests of algorithms where we want all that to be abstracted away. In fact, I think we should probably wrap *all* the OpenGL stuff we do for ant_world_db_creator into ``AntAgent``, so that people like me who don't understand OpenGL don't have to worry about it :-).

I agree that it would be nice to have a generic way to access the cool visualisations that you have for ant_world and stone_cx, but we don't have to try to implement that now. It totally makes sense to have libantworld as a lower-level API that gives us the extra flexibility if we want it. Longer term we could write a generic interface for giving us the top-down view in parallel etc. (maybe by inheriting from ``AntAgent``).

Reply:
No, that is fair enough!